src/kernels/helpers.ts

```diff
@@ -246,9 +251,87 @@ export function findPreferredKernel(
     } else {
         traceInfoIfCI(`Preferred kernel is ${JSON.stringify(preferredKernel)}`);
     }
+
     return preferredKernel;
 }
 
+function isExactMatch(
+    kernelConnection: KernelConnectionMetadata,
+    notebookMetadata: nbformat.INotebookMetadata
+): boolean {
+    // To get an exact match, we need to have a kernelspec in the metadata
+    if (!notebookMetadata.kernelspec) {
+        return false;
+    }
+
+    // Live kernel connections are not exact matches
+    if (kernelConnection.kind === 'connectToLiveRemoteKernel') {
+        return false;
+    }
+
+    if (notebookMetadata.interpreter && interpreterMatchesThatInNotebookMetadata(kernelConnection, notebookMetadata)) {
+        // Case: Metadata has interpreter, in this case it should have an interpreter
+        // and a kernel spec that should fully match, note that in this case matching
+        // name on a default python kernel spec is ok (as the interpreter hash matches)
+        return isKernelSpecExactMatch(kernelConnection, notebookMetadata.kernelspec, true);
+    } else {
+        // Case: Metadata does not have an interpreter, in this case just full match on the
+        // kernelspec, but do not accept default python name as valid for an exact match
+        return isKernelSpecExactMatch(kernelConnection, notebookMetadata.kernelspec, false);
+    }
+}
+
+function isKernelSpecExactMatch(
+    kernelConnection: KernelConnectionMetadata,
+    notebookMetadataKernelSpec: nbformat.IKernelspecMetadata,
+    allowPythonDefaultMatch: boolean
+): boolean {
+    if (kernelConnection.kind === 'connectToLiveRemoteKernel') {
+        return false;
+    }
+    const kernelConnectionKernelSpec = kernelConnection.kernelSpec;
+
+    // Get our correct kernelspec name and display_name from the connection
+    const connectionOriginalSpecFile =
+        kernelConnectionKernelSpec.metadata?.vscode?.originalSpecFile ||
+        kernelConnectionKernelSpec.metadata?.originalSpecFile;
+    const connectionKernelSpecName = connectionOriginalSpecFile
+        ? path.basename(path.dirname(connectionOriginalSpecFile))
+        : kernelConnectionKernelSpec?.name || '';
+    const connectionKernelSpecDisplayName =
+        kernelConnectionKernelSpec.metadata?.vscode?.originalDisplayName ||
+        kernelConnectionKernelSpec.display_name ||
+        '';
+    const connectionInterpreterEnvName = kernelConnection.interpreter?.envName;
+
+    if (
+        allowPythonDefaultMatch &&
+        connectionKernelSpecDisplayName === notebookMetadataKernelSpec.display_name &&
+        (connectionKernelSpecName === notebookMetadataKernelSpec.name ||
```

Comment:
Mimicking compareKernels check for either name or the envName of the kernelConnection interpreter.